docs/design/storage-implementation.md

```diff
@@ -0,0 +1,157 @@
+- Training jobs of experiment typically generate model artifacts (files) which need perisisted, and both of notebook, model serving needs to load model artifacts from persistent storage. 
+- There're various of meta information, such as experiment meta, model registry, model serving, notebook, experiment, environment, etc. We need be able to read these meta information back.
+- We also have code for experiment (like training/batch-prediction), notebook (ipynb), and endpoints.
+- And notebook/experiments/endpoint need depend on environments (dependencies such as pip, and Docker Images).
+
+### Implementation considerations for ML-related objects
+
+| Object Type                              | Characteristics                                              | Where to store                                               |
+| ---------------------------------------- | ------------------------------------------------------------ | ------------------------------------------------------------ |
+| Metrics: tf.events                       | Time series data with k/v, appendable to file                | Local/EBS, HDFS, Cloud Blob Storage                          |
+| Metrics: other tracking metrics          | Time series data with k/v, appendable to file                | Local, HDFS, Cloud Blob Storage, Database                    |
+| Logs                                     | Large volumes, #files are potentially huge.                  | Local (temporary), HDFS (need aggregation), Cloud Blob Storage |
+| Submarine Metastore                      | CRUD operations for small meta data.                         | Database                                                     |
+| Model Artifacts                          | Size varies for model (from KBs to GBs). #files are potentially huge. | HDFS, Cloud Blob Storage                                     |
+| Code                                     | Need version control. (Please find detailed discussions below for code storage and localization) | Tarball on HDFS/Cloud Blog Storage, or Git                   |
+| Environment (Dependencies, Docker Image) |                                                              | Public/private environment repo (like Conda channel), Docker registry. |
+
+### Detailed discussions
+
+#### Store code for experiment/notebook/model-serving
+
+There're following ways to get experiment code: 
+
+**1) Code is part of Docker image:** 
+
+***This is an anti-pattern and we will NOT recommend you to use it***, Docker image can be used to include ANYTHING, like dependencies, the code you will execute, or even data. But this doesn't mean you should do it. We recommend to use Docker image ONLY for libraries/dependencies.
+
+Making code to be part of Docker image makes hard to edit code (if you want to update a value in your Python file, you will have to recreate the Docker image, push it and rerun it).
+
+**2) Code is part of Git repo:** 
```

Comment:
Because this is our recommended way, it is better to put it on the first line.
**1) Code is part of Git repo:** (recommended)

Reply:
Make sense, updated.